### ISSUE TITLE
Support for different default language

### ISSUE THREAD

**Issue by jcolladosp** [microsoft/vscode#168127]

**Comment by deakjahn** [comment]
@TylerLeonhardt I hope you don't forget that the other part of this equation is to be able to refer to _all_ strings with a key rather than the default locale string itself. This is how most of the various platforms solve the question of localization, even if it means an extra lookup (with a proper hash dictionary, with today's computers, it's hardly any real problem).

**Comment by jcolladosp** [comment]
Hi,
any update on this?
Thanks

**Comment by wszgrcy** [comment]
I translate by myself.......
```ts
import * as vscode from 'vscode';
import * as path from 'path';
import * as fs from 'fs';
let cache: Record<string, string>;
export function $translate(key: string, obj?: Record<string, any>) {
  if (!cache) {
    let filePath = path.join(__dirname, `./l10n/bundle.l10n.${vscode.env.language}.json`);
    if (!fs.existsSync(filePath)) {
      cache = __non_webpack_require__(`./l10n/bundle.l10n.json`);
    } else {
      cache = __non_webpack_require__(filePath);
    }
  }
  let value = cache[key] ?? key;
  if (!obj) {
    return value;
  }

  for (const key in obj) {
    value = value.replace(new RegExp(`{\s*${key}\s*}`, 'g'), obj[key]);
  }
  return value;
}

```

**Comment by roco1234** [comment]
Any update on this?

**Comment by yiliang114** [comment]
> This would need to be supported in Core. Transferring there.这需要在 Core 中得到支持。转移到那里。

@TylerLeonhardt  It seems that no one has followed up on this issue, I want to try to do it.

**Comment by TylerLeonhardt** [comment]
@yiliang114 I don't think this will be an easy one... but I'm open to discussing how we can achieve this.

**Comment by wdhongtw** [comment]
> I think a better solution to this problem is to allow some override setting in the l10n property of the manifest that says:

```
{
  "l10n": {
    "location":"same as just specifying l10n as a string",
    "defaultLocale": "es"
  }
}
```

> I don't think this will be an easy one... but I'm open to discussing how we can achieve this.

I have a sample implementation following the proposal above, and it seems to works.
(At least works during my manual test.)

We can adjust `ExtHostLocalizationService` for three things:

- `initializeLocalizedMessages`: check if current language be matched with extension default language, cache the result, and skip bundle loading step accordingly.
- `getBundleLocation`: adjust manifest parsing logic to support two kinds of schema for l10n path.
- `getMessage`: check if current translation operation is required, based on the per-extension cache

So if any extension has extension-default-language matched with current vscode language, no additional file is loaded.
And to be backward compatible, extension-default-language fallbacks to `LANGUAGE_DEFAULT` (en).

There should be no problem for loading step, just constant time overhead.
For actual translation `getMessage`, the overhead is one map lookup with string key.
Assuming that extension ID operation can be benifit from short string optimization, it's probably ok.

See: https://github.com/wdhongtw/vscode/commit/d7d63ad3411ec0d548fb74341c7be3d7e7f53e24

@TylerLeonhardt 

Is this looks good to you? If all goes well, maybe I can start rebasing and submit a PR for formal review process.

**Comment by TylerLeonhardt** [comment]
@wdhongtw definitely interested. Keep in mind, that an extension that sets this needs a mechanism to read in their English strings from `bundle.l10n.en.json`. Also, there are tooling implications which would be work in:
https://github.com/microsoft/vscode-l10n/tree/main/l10n-dev

That assumes bundle.l10n.json is English, for example, so that would need to be handled.

**Comment by TylerLeonhardt** [comment]
Sounds great!
